ROOTFS/usr/bin/waggle-network-watchdog

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python3
+from pathlib import Path
+import subprocess
+import time
+import logging
+from glob import glob
+
+logging.basicConfig(level=logging.INFO)
+
+last_connection_time = 0
+
+
+def update_systemd_watchdog():
+    try:
+        subprocess.check_call(["systemd-notify", "WATCHDOG=1"])
+    except subprocess.CalledProcessError:
+        logging.warning("skipping reset of systemd watchdog")
+
+
+def time_now():
+    return time.monotonic()
+
+
+def seconds_since(start):
+    return time.monotonic() - start
+
+
+def ssh_connection_ok():
+    try:
+        return (
+            "beehive:20022"
+            in subprocess.check_output(["ss", "-tr", "state", "established"]).decode()
+        )
+    except Exception:
+        return False
+
+
+# NOTE Revisit how much of the network stack we should restart. For now, I want to cover all
+# cases of wifi and modems and ssh tunnel issues.
+def restart_network_services():
+    logging.warning("restarting network services")
+
+    # ensure proper ownership of ports, ttyACM* for Modem
+    ports = glob("/dev/ttyACM*")
+    subprocess.run(["chown", "root:root"] + ports)
+    subprocess.run(["chmod", "660"] + ports)
+
+    # restart network services
+    subprocess.run(
+        [
+            "systemctl",
+            "restart",
+            "NetworkManager",
+            "ModemManager",
+            "waggle-reverse-tunnel",
+        ]
+    )
+
+
+def reboot_os():
+    logging.warning("rebooting the system")
+    # aggressively but safely reboot the system
+    subprocess.run(["systemctl", "--force", "reboot"])
+
+
+last_connection_time = time_now()
+
+# Recovery actions table [time (s), recovery function]
+# restart networking stack after 15, 20 and 25 of no beehive connectivity
+# reboot after 30 mins of no beehive connectivity
+recovery_actions = [
+    (1800, reboot_os),
+    (1500, restart_network_services),
+    (1200, restart_network_services),
+    (900, restart_network_services),
+]
+
+while True:
+    update_systemd_watchdog()
+
+    if ssh_connection_ok():
+        logging.info("connection ok")
+        last_connection_time = time_now()
+    else:
+        logging.warning(
+            "no connection for %ss", int(seconds_since(last_connection_time))
+        )
+
+    for action in recovery_actions:
+        if seconds_since(last_connection_time) >= action[0]:
```

Comment:
I don't see any kind of cool down on triggering this action. It looks like this will just keep retriggering the action once the time since connection is above any of the thresholds.